agent/lib_aws_sdk_php.c

```diff
@@ -295,6 +302,166 @@ void nr_lib_aws_sdk_php_sqs_parse_queueurl(
+
+  /* Determine if we instrument this command. */
+  if (AWS_COMMAND_IS("invoke")) {
+    // Command can be saved here if in the future
+    // we instrument more than 1 lambda command
+  } else {
+    return;
+  }
+#undef AWS_COMMAND_IS
+
+  /* reconstruct the ARN */
+  nr_aws_lambda_invoke(NR_EXECUTE_ORIG_ARGS, &cloud_attrs);
+  if (!cloud_attrs.cloud_resource_id) {
+    /* we do not want to instrument if we cannot reconstruct the ARN */
+    return;
+  }
+
+  /*
+   * By this point, it's been determined that this call will be instrumented so
+   * only create the segment now, grab the parent segment start time, add our
+   * special segment attributes/metrics then close the newly created segment.
+   */
+  external_segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+  if (NULL == external_segment) {
+    return;
+  }
+  /* re-use start time from auto_segment started in func_begin */
+  external_segment->start_time = auto_segment->start_time;
+  cloud_attrs.aws_operation = command_name_string;
+
+  /* end the segment */
+  nr_segment_traces_add_cloud_attributes(external_segment, &cloud_attrs);
+  nr_segment_external_params_t external_params = {.library = "aws_sdk"};
+  zval* data = nr_php_get_zval_object_property(*retval_ptr, "data");
+  if (NULL != data && IS_ARRAY == Z_TYPE_P(data)) {
+    zval* status_code = nr_php_zend_hash_find(Z_ARRVAL_P(data), "StatusCode");
+    if (NULL != status_code && IS_LONG == Z_TYPE_P(status_code)) {
+      external_params.status = Z_LVAL_P(status_code);
+    }
+    zval* metadata = nr_php_zend_hash_find(Z_ARRVAL_P(data), "@metadata");
+    if (NULL != metadata  && IS_REFERENCE == Z_TYPE_P(metadata)
+        && IS_ARRAY == Z_TYPE_P(Z_REFVAL_P(metadata))) {
+      zval* uri = nr_php_zend_hash_find(Z_ARRVAL_P(Z_REFVAL_P(metadata)), "effectiveUri");
+      if (NULL != uri && IS_STRING == Z_TYPE_P(uri)) {
+        external_params.uri = Z_STRVAL_P(uri);
+      }
+    }
+  }
+  nr_segment_external_end(&auto_segment, &external_params);
+}
+
+void nr_aws_lambda_invoke(NR_EXECUTE_PROTO, nr_segment_cloud_attrs_t* cloud_attrs) {
+  zval* call_args = nr_php_get_user_func_arg(2, NR_EXECUTE_ORIG_ARGS);
+  zval* this_obj = NR_PHP_USER_FN_THIS();
+  char* arn = NULL;
+  char* function_name = NULL;
+  char* region = NULL;
+  zval* region_zval = NULL;
+  char* qualifier = NULL;
+  zval* qualifier_zval = NULL;
+  char* accountID = NULL;
+
+  /* verify arguments */
+  if (NULL == call_args || IS_ARRAY != Z_TYPE_P(call_args)) {
+    return;
+  }
+  zval* lambda_args = nr_php_zend_hash_index_find(Z_ARRVAL_P(call_args), 0);
+  if (NULL == lambda_args || IS_ARRAY != Z_TYPE_P(lambda_args)) {
+    return;
+  }
+  zval* lambda_name = nr_php_zend_hash_find(Z_ARRVAL_P(lambda_args), "FunctionName");
+  if (NULL == lambda_name || IS_STRING != Z_TYPE_P(lambda_name)) {
+    return;
+  }
+
+  /* Compile the regex */
+  if (NULL == NRPRG(aws_arn_regex)) {
+    NRPRG(aws_arn_regex) = nr_regex_create(AWS_ARN_REGEX, 0, 0);
```

Comment:
Where/when does this get destroyed?

Reply:
Can we just precompile this in minit like wordpress does?
https://github.com/newrelic/newrelic-php-agent/blob/main/agent/fw_wordpress.c#L878-L886

Reply:
This has been moved to be compiled on first use and destroyed on mshutdown.

Reply:
Instead of first use (which would make the first instrumented call take much longer than any other call, why not have it setup either in minit or in nr_aws_sdk_php_enable if you want to wait until at least the agent knows aws-sdk is involved.

Reply:
In an effort to minimize the number of times that this regex is unnecessarily compiled, I would like to keep the compilation where it is.